### DISCUSSION TITLE
Standard library and Pydantic v2 dataclasses: ResponseValidationError after upgrade

### DISCUSSION THREAD

**Discussion by Hellzed** [fastapi/fastapi#10427] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import dataclasses

import pydantic
from fastapi import FastAPI

app = FastAPI()


@dataclasses.dataclass
class CustomModel:
    message: str


@pydantic.dataclasses.dataclass
class OutputModel:
    message: str


@app.get("/", response_model=OutputModel)
async def root() -> CustomModel:
    return CustomModel(message="Hello World")
```

### Description

I'm in the process of migrating a codebase to Pydantic v2.

Services return standard library dataclasses, which are validated and publicly documented by Pydantic dataclasses, as per the example above.
Pydantic validation can't be added to the service dataclasses themselves because it would strip third party instrumentation, which is not desirable.

After upgrading from Pydantic 1.10.13 to Pydantic 2.4.2, calling any route that returns a standard library dataclass but declares a Pydantic dataclass `response_model` will fail with a `fastapi.exceptions.ResponseValidationError`.

Allowing validation of arbitrary models (formerly known as "ORM mode"), as `@pydantic.dataclasses.dataclass(config=pydantic.ConfigDict(from_attributes=True))`, doesn't have any effect.

Using `response_model=None` will output the fully serialized `CustomModel` instance, which is not the intended behavior (we need documentation and output filtering).

Am I doing something wrong? Is there a straightforward way to support this?

### Operating System

Linux

### Operating System Details

Ubuntu 23.04 (Docker)

### FastAPI Version

0.103.0

### Pydantic Version

2.4.2

### Python Version

3.11.4

### Additional Context

Full error output:
```fastapi.exceptions.ResponseValidationError: 1 validation errors:
  {
   "type":"dataclass_type",
   "loc":"(""response",
   ")",
   "msg":"Input should be a dictionary or an instance of OutputModel",
   "input":"CustomModel(message=""Hello World"")",
   "ctx":{
      "class_name":"OutputModel"
   },
   "url":"https://errors.pydantic.dev/2.4/v/dataclass_type"
}```

**Comment by Hellzed** [comment]
This might actually be a Pydantic issue:
#https://github.com/pydantic/pydantic/issues/7805

**Comment by YuriiMotov** [comment, marked as answer]
It's said in the [comment to the issue on Pydantic](https://github.com/pydantic/pydantic/issues/7805#issuecomment-1759586884) that pydantic dataclass doesn't support `from_attributes` and this is intentionally.

So, the only way to go is to use `BaseModel`
```python
class OutputModel(pydantic.BaseModel):
    message: str
```
